scripts/gn/parse_rinex_header.py

```diff
@@ -0,0 +1,186 @@
+"""
+This script parses info from a RINEX v3 obs file. This info is used
+downstream to:
+- determine what needs to be downloaded for Ginan to be able to run.
+"""
+from datetime import datetime, timedelta, date
+from pathlib import Path
+from typing import Tuple, NamedTuple, Optional, Dict
+import re
+
+import georinex as gr
+import xarray
+
+# Mappings from character code to Ginan name
+GNSS_SYSTEMS = {"G": "gps", "R": "glo", "S": "sbas", "C": "bds", "E": "gal", "J": "qzs"}
+
+
+class Receiver(NamedTuple):
+    number: str
+    type: str
+    version: str
+
+
+class AntennaDeltas(NamedTuple):
+    height: float
+    north: float
+    east: float
+
+
+class Antenna(NamedTuple):
+    type: str
+    deltas: AntennaDeltas
+
+    def get_eccentricity(self):
+        """Ginan expects ENU, but RINEX is UNE"""
+        deltas = self.deltas
+        return [deltas.east, deltas.north, deltas.height]
+
+
+class ApproxPosition(NamedTuple):
+    x: float
+    y: float
+    z: float
+
+    def get_apriori_position(self):
+        return [self.x, self.y, self.z]
+
+
+class RinexHeader(NamedTuple):
+    marker_name: str
+    receiver: Receiver
+    antenna: Antenna
+    approx_position: ApproxPosition
+    first_obs_time: datetime
+    last_obs_time: Optional[datetime]
+    sys_signals: Dict[str, str]
+
+    def get_station_alias(self):
+        """Ginan expects a four character string for each station. A station may have more characters than this,
+        so we need to create an alias."""
+        alias = self.marker_name
+        if len(self.marker_name) > 4:
+            # Rename the rinex file to RXXX.rnx, where XXX are the last 3 characters of the marker name.
+            alias = f"R{self.marker_name[-3:]}"
+        return alias
+
+
+def parse_v3_header(filepath: Path) -> RinexHeader:
+    header = gr.rinexheader(filepath)
+    if int(header["version"]) != 3:
+        raise NotImplementedError("Only RINEX v3 is currently supported")
+
+    # Load observations to determine which signals have been observed
+    # by the receiver for each gnss system (GPS, Galileo etc)
+    obs = gr.load(filepath)
+    sys_signals = _get_signals_per_system(obs)
+
+    marker_name = header["MARKER NAME"].strip()
+
+    # TODO: Handle the case when not all the fields are given - can't unpack 3 fields when there are only two values
```

Comment:
I appreciate the call out on this one, given issues I've seen before.
Can we determine in a straightforward way whether the fields are missing, and raise an exception on it?

Reply:
It's hard to tell which fields are missing. It's possible that only the TYPE (middle value) could be provided, and when we split we would only get one value back. It would be hard to tell that we're missing "REC #" and "VERS".

We would need some kind of way to check that each element matches a pattern, ie. check if the TYPE is in a list of valid receivers, does the REC # conform to a certain pattern, is the VERS valid. These would be assumptions and I'm not sure if user data would always follow these patterns.

I'm not sure the best way of handling it at the moment.

Reply:
Right. Yeah hard to handle if less than three values **could be valid**, but telling what is what, and validating it is non-trivial.

I guess !=3 elements: raise `NotImplemented` for now? (Are we actually looking at values though, or just the header?)

I can't quite see how the `TYPE` is being _extracted_ as it is on line 82.
~~Line 69 is also confusing me... ohh~~. I'd suggest fully qualifying `gr` as `georinex`. I misread it as `gn` several times and couldn't figure how anything was working given `rinexheader()` isn't defined locally, but there's a class called `RinexHeader` with no explicit constructor.

Reply:
Line 82 was what I was using to parse a specific file that I have that only has TYPE (not the other two elements), but it needs to be more general.

The only element that's being used is TYPE (receiver type), which ends up in `receiver_options: STATION: receiver_type:` in the ginan yaml. Does ginan actually use this information for anything? Or does it just care about the *antenna type* to pull out the correct values from the igs20.atx file?

If it's not using it, then perhaps we don't parse it at all for now? If it is, then could we split the field into its elements and check any of the elements are found in a list of valid TYPEs (not sure where to get this list but I'm sure it would exist somewhere).

It would be nice if we could raise an exception if there are not 3 elements, but I've come across a fair few files that don't have this field fully populated. Trying our best to pull out a valid type might be nicer than asking the user to pad out the field with dummy data so that there are 3 elements to parse.

Reply:
Apparently `receiver_type` hardly gets used. And I haven't found a list of valid `TYPE`s anywhere.
So given it's also hard to parse out robustly, I'd agree we probably shouldn't parse it for now.

Reply:
See [4cedd91](https://github.com/GeoscienceAustralia/ginan/pull/62/commits/4cedd911bd985f36f8690452dbcca2a10d8f0941)